allennlp_models/seq2seq/bart/bart.py

```diff
@@ -0,0 +1,401 @@
+            initial_decoder_ids = torch.cat(
+                [
+                    torch.full(
+                        (input_ids.shape[0], 1),
+                        self._decoder_start_id,
+                        dtype=input_ids.dtype,
+                        device=input_ids.device,
+                    ),
+                    torch.full(
+                        (input_ids.shape[0], 1),
+                        self._start_id,
+                        dtype=input_ids.dtype,
+                        device=input_ids.device,
+                    ),
+                ],
+                dim=-1,
+            )
+
+            inital_state = {
+                "input_ids": input_ids,
+                "input_mask": input_mask,
+                "encoder_states": None,
+            }
+            beam_result = self._beam_search.search(
+                initial_decoder_ids, inital_state, self.take_step
+            )
+
+            predictions = beam_result[0]
+            max_pred_indices = (
+                beam_result[1].argmax(dim=-1).view(-1, 1, 1).expand(-1, -1, predictions.shape[-1])
+            )
+            predictions = predictions.gather(dim=1, index=max_pred_indices).squeeze(dim=1)
+
+            self._rogue(predictions, target_ids)
+            self._bleu(predictions, target_ids)
+
+            outputs["predictions"] = predictions
+            outputs["log_probabilities"] = (
+                beam_result[1].gather(dim=-1, index=max_pred_indices[..., 0]).squeeze(dim=-1)
+            )
+
+            self.make_output_human_readable(outputs)
+
+        return outputs
+
+    @staticmethod
+    def _decoder_cache_to_dict(decoder_cache):
+        cache_dict = {}
+        for layer_index, layer_cache in enumerate(decoder_cache):
+            for attention_name, attention_cache in layer_cache.items():
+                for tensor_name, cache_value in attention_cache.items():
+                    key = (layer_index, attention_name, tensor_name)
+                    cache_dict[key] = cache_value
+        return cache_dict
+
+    @staticmethod
+    def _dict_to_decoder_cache(cache_dict):
+        decoder_cache = []
+        for key, cache_value in cache_dict.items():
+            # Split key and extract index and dict keys
+            layer_idx, attention_name, tensor_name = key
+            # Extend decoder_cache to fit layer_idx + 1 layers
+            decoder_cache = decoder_cache + [{} for _ in range(layer_idx + 1 - len(decoder_cache))]
+            cache = decoder_cache[layer_idx]
+            if attention_name not in cache:
+                cache[attention_name] = {}
+            assert tensor_name not in cache[attention_name]
+            cache[attention_name][tensor_name] = cache_value
+        return decoder_cache
+
+    def take_step(
+        self, last_predictions: torch.Tensor, state: Dict[str, torch.Tensor], step: int
+    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
```

Comment:
Do you have to turn off gradients somewhere in here, or is that taken care of outside?

Reply:
Good catch, this doesn't seem to be done anywhere, also not for the other seq2seq classes that use our beam search code. I added a `torch.no_grad` to the beam search code so that we disable it in all of those cases. I'll push this to the bart branch on allennlp.